kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -804,13 +842,184 @@ impl<'tcx> GotocCtx<'tcx> {
+                }
+            };
+            let mut fields = vec![];
+            // generate a struct for the direct fields of the layout (fields that don't occur in the variants)
+            fields.push(DatatypeComponent::Field {
+                name: "direct_fields".into(),
+                typ: ctx.codegen_generator_variant_struct(
+                    generator_name,
+                    pretty_name,
+                    type_and_layout,
+                    "DirectFields".into(),
+                    Some(*discriminant_field),
+                ),
+            });
+            for var_idx in variants.indices() {
+                let variant_name = GeneratorSubsts::variant_name(var_idx).into();
+                fields.push(DatatypeComponent::Field {
+                    name: ctx.generator_variant_name(var_idx),
+                    typ: ctx.codegen_generator_variant_struct(
+                        generator_name,
+                        pretty_name,
+                        type_and_layout.for_variant(ctx, var_idx),
+                        variant_name,
+                        None,
+                    ),
+                });
+            }
+            fields
+        });
+        generator_enum
+    }
+
+    fn codegen_generator_variant_struct(
+        &mut self,
+        generator_name: InternedString,
+        pretty_generator_name: InternedString,
+        type_and_layout: TyAndLayout<'tcx, Ty<'tcx>>,
+        variant_name: InternedString,
+        discriminant_field: Option<usize>,
+    ) -> Type {
+        let struct_name = format!("{generator_name}::{variant_name}");
+        let pretty_struct_name = format!("{pretty_generator_name}::{variant_name}");
+        debug!("codegen generator variant struct {}", pretty_struct_name);
+        self.ensure_struct(struct_name, pretty_struct_name, |ctx, _| {
+            let mut offset = Size::ZERO;
+            let mut fields = vec![];
+            for idx in type_and_layout.fields.index_by_increasing_offset() {
+                let field_name = if Some(idx) == discriminant_field {
+                    "case".into()
+                } else {
+                    ctx.generator_field_name(idx).into()
+                };
+                let field_ty = type_and_layout.field(ctx, idx).ty;
+                let field_offset = type_and_layout.fields.offset(idx);
+                let field_size = type_and_layout.field(ctx, idx).size;
+                if let Some(padding) =
+                    ctx.codegen_struct_padding(offset.bits(), field_offset.bits(), idx)
+                {
+                    fields.push(padding);
+                }
+                fields.push(DatatypeComponent::Field {
+                    name: field_name,
+                    typ: ctx.codegen_ty(field_ty),
+                });
+                offset += field_size;
+            }
+            // TODO: is this necessary? I don't see how padding at the end matters. But codegen_struct_fields does it too.
```

Comment:
@danielsn @celinval Do you know if/why this part of the code is necessary? I copied it from `codegen_struct_fields`. Intuitively, it seems superfluous to me.

Reply:
I did some experiments: if i remove it from the generator code, all tests still pass. (But I don't know whether the tests just don't cover the necessary case?) If I remove it from `codegen_struct_fields`, tests fail.

Reply:
Just discussed with @danielsn. The code is important to get the right sizeof. I'll remove the TODO.